src/Storages/StorageS3.cpp

```diff
@@ -125,6 +127,138 @@ namespace ErrorCodes
+        {
+            auto column = column_type->createColumn();
+            for (const auto & key : keys)
+            {
+                auto pos = key.find_last_of('/');
+                if (pos != std::string::npos)
+                    column->insert(key.substr(pos + 1));
+                else
+                    column->insert(key);
+            }
+            virtual_columns_block.insert({std::move(column), column_type, column_name});
+        }
+        else if (column_name == "_key")
+        {
+            auto column = column_type->createColumn();
+            for (const auto & key : keys)
+                column->insert(key);
+            virtual_columns_block.insert({std::move(column), column_type, column_name});
+        }
+        else
+        {
+            auto column = column_type->createColumn();
+            column->insertManyDefaults(keys.size());
+            virtual_columns_block.insert({std::move(column), column_type, column_name});
+        }
+    }
+
+    /// Column _key is mandatory and may not be in virtual_columns list
+    if (!virtual_columns_block.has("_key"))
+    {
+        auto column_type = std::make_shared<DataTypeString>();
+        auto column = column_type->createColumn(); for (const auto & key : keys)
+            column->insert(key);
+        virtual_columns_block.insert({std::move(column), column_type, "_key"});
+    }
+
+    return virtual_columns_block;
+}
+
+static Block renameColumnsInBlock(const Block & source_block, const std::unordered_map<String, String> & rename_map)
+{
+    auto columns = source_block.getColumnsWithTypeAndName();
+    for (auto & col : columns)
+    {
+        auto it = rename_map.find(col.name);
+        if (it != rename_map.end())
+            col.name = it->second;
+    }
+    return Block(std::move(columns));
+}
+
+static ActionsDAGPtr getFilterForPartitionPruning(const SelectQueryInfo & query_info,
+                                                  const NamesAndTypesList & virtual_columns,
+                                                  NameToNameMap & column_rename,
+                                                  ContextPtr context)
+{
+    if (!query_info.query_tree || !query_info.planner_context)
+        return nullptr;
+
+    const auto * query_node = query_info.query_tree->as<QueryNode>();
+    if (!query_node || !query_node->getWhere())
+        return nullptr;
+
+    Block header = getBlockWithVirtuals(virtual_columns, "", {});
+
+    const auto & table_expression_data = query_info.planner_context->getTableExpressionDataOrThrow(query_info.table_expression);
+    for (const auto & [column_name, column_identifier] : table_expression_data.getColumnNameToIdentifier())
+        column_rename.emplace(column_name, column_identifier);
+
+    header = renameColumnsInBlock(header, column_rename);
+    auto filter_dag = buildActionsDAGFromExpressionNode(
+        query_node->getWhere(), header.getColumnsWithTypeAndName(), query_info.planner_context);
```

Comment:
This looks very suspicious. You should not extract filter from the query tree. Filter must be stored in the `SelectQueryInfo::filter_actions_dag`.
`PartitionPruner` uses `SelectQueryInfo::filter_actions_dag`.